### PR TITLE
DGS-4220: fix request tag based metrics

### PR DESCRIPTION
**Pull request by ericwush**
This PR tries to fix the bug that a sensor created for a method + performance annotation pair is shared across the entire application, as opposed to each request tag set having its own copy of sensor. It is causing a problem when request tags are used. A sensor is either created or fetched depending on whether the name exists. For example, 
1. `methodX.request-size` sensor is created on application startup, let's call it sensorA
2. a request w/ tag1 comes in, sensorA is returned because the same sensor name is provided
3. a request w/ tag2 comes in, sensorA is again returned

However, each tag really should have its own sensor.

The fix sorts request tags to make sure the same sensor is used for the same set of request tags. Tags in `MetricName` is also sorted for `JmxReporter` to create consistent MBean names. I understand this is an implicit side effect, a more proper place to ensure the consistency is within `JmxReporter`.

This PR also fixes a bug that the error sensor is overwritten, so error rates are not recorded.

### CODE REVIEW

**Comment by ehumber** [comment]
@codyaray Is this something your team can look at reviewing?

**Comment by ericwush** [comment]
@rayokota should be able to help with the review.

BTW, I have done a code search, looks like only our team is using request tags by setting `REQUEST_TAGS_PROP_KEY`. Just double check @ehumber your team doesn't use it, otherwise, the according alerts might need to be adjusted.

File: core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java
Comment (rayokota) [inline core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java:150]: Can this be replaced with `computeIfAbsent`?

**Review by rayokota — APPROVED** [review]
One comment, otherwise LGTM

**Comment by ehumber** [comment]
> @rayokota should be able to help with the review.
> 
> BTW, I have done a code search, looks like only our team is using request tags by setting `REQUEST_TAGS_PROP_KEY`. Just double check @ehumber your team doesn't use it, otherwise, the according alerts might need to be adjusted.

I've looked, I can't see any usage of `REQUEST_TAGS_PROP_KEY` in the kafka-rest or ce-kafka-rest code.

**Comment by ericwush** [comment]
Thanks @ehumber. Seems I need the code owner's approval to merge. Otherwise, could you merge it for me?

**Comment by codyaray** [comment]
cc @rigelbm

File: core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java
Comment (ehumber) [inline core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java:194]: When making a similar change in kafka-rest, I was asked to ensure the sensors expire if they are unused, otherwise we will end up with a sensor leak.

1 hour was the suggested duration.

This didn't matter so much when one sensor was reused, but now there are multiple sensors I think it is important.

**Review by ehumber — APPROVED** [review]

**Comment by ehumber** [comment]
@ericwush Approved but with one comment that needs addressing before you merge.

File: core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java
Comment (ericwush) [reply core/src/main/java/io/confluent/rest/metrics/MetricsResourceMethodApplicationListener.java:195]: Good point, addressed, thanks!

**Comment by ericwush** [comment]
Thanks @ehumber for the review. As discussed, I've addressed the feedback and also included the bug fix for error sensor.

**Comment by ehumber** [comment]
LGTM

**Pull request merged**
